shared/bazel/rules/robotpy/semiwrap_helpers.bzl

```diff
@@ -0,0 +1,345 @@
+        outs = [output_json, output_pc],
+        cmd = cmd,
+        tools = _wrapper_dep() + typecasters_srcs + [package_root],
+        visibility = ["//visibility:public"],
+        tags = ["robotpy"],
+    )
+
+def resolve_casters(
+        name,
+        casters_pkl_file,
+        dep_file,
+        caster_files = [],
+        caster_deps = []):
+    """
+    Sugar wrapper for the semiwrap.cmd.resolve_casters tool
+    """
+    cmd = _wrapper() + " semiwrap.cmd.resolve_casters "
+    cmd += " $(OUTS)"
+
+    cmd += _location_helper(_semiwrap_caster()) + "/semiwrap/semiwrap.pybind11.json"
+
+    resolved_caster_files = []
+
+    deps = []
+    for dep in caster_deps:
+        deps.append(dep)
+        cmd += _location_helper(dep)
+
+    for cfd in caster_files:
+        if cfd.startswith(":"):
+            resolved_caster_files.append(cfd)
+            cmd += _location_helper(cfd)
+        else:
+            cmd += " " + cfd
+
+    native.genrule(
+        name = name,
+        srcs = resolved_caster_files + deps,
+        outs = [RESOLVE_CASTERS_DIR + casters_pkl_file, RESOLVE_CASTERS_DIR + dep_file],
+        cmd = cmd,
+        tools = _wrapper_dep() + [_semiwrap_caster()],
+        tags = ["robotpy"],
+    )
+
+def gen_libinit(
+        name,
+        output_file,
+        modules):
+    """
+    Sugar wrapper for the semiwrap.cmd.gen_libinit tool
+    """
+    cmd = _wrapper() + " semiwrap.cmd.gen_libinit "
+    cmd += " $(OUTS) "
+    cmd += " ".join(modules)
+
+    native.genrule(
+        name = name,
+        outs = [output_file],
+        cmd = cmd,
+        tools = _wrapper_dep(),
+        tags = ["robotpy"],
+    )
+
+def gen_pkgconf(
+        name,
+        project_file,
+        module_pkg_name,
+        pkg_name,
+        output_file,
+        libinit_py,
+        install_path,
+        package_root):
+    """
+    Sugar wrapper for the semiwrap.cmd.gen_pkgconf tool
+    """
+    cmd = _wrapper() + " semiwrap.cmd.gen_pkgconf "
+    cmd += " " + module_pkg_name + " " + pkg_name
+    cmd += _location_helper(project_file)
+    cmd += " $(OUTS)"
+    if libinit_py:
+        cmd += " --libinit-py " + libinit_py
+
+    OUT_FILE = install_path + "/" + output_file
+    native.genrule(
+        name = name,
+        srcs = [package_root],
+        outs = [OUT_FILE],
+        cmd = cmd,
+        tools = _wrapper_dep() + [project_file],
+        visibility = ["//visibility:public"],
+        tags = ["robotpy"],
+    )
+
+def header_to_dat(
+        name,
+        casters_pickle,
+        include_root,
+        class_name,
+        yml_file,
+        header_location,
+        generation_includes = [],
+        header_to_dat_deps = [],
+        extra_defines = [],
+        deps = []):
+    """
+    Sugar wrapper for the semiwrap.cmd.header2dat tool
+    """
+    cmd = _wrapper() + " semiwrap.cmd.header2dat "
+    cmd += "--cpp 202002L "  # TODO(pj) - This is the option when I ran on linux. Does its value really matter?
```

Comment:
The value does matter as it varies depending on what C++ version you're building with.

The `makeplan` output it emits `CppMacroValue("__cplusplus")` to determine this value, and so you should resolve it at build time and plug it in to this command.

Reply:
See how I do it for meson at https://github.com/robotpy/semiwrap/blob/83ffd5e0158d7ecabe2c27f5a81521a87657a17d/src/semiwrap/render_meson.py#L359

Reply:
When cxxheaderparser parses the header files, it needs a C++ preprocessor to process them first. If it uses gcc to do preprocessing (default on Linux/OSX) then it will already set the value of the `__cplusplus` macro because it's builtin. On Windows, we can't use the preprocessor for MSVC, so it uses `pcpp` instead which doesn't have a `__cplusplus` macro defined.

The reason the macro matters is there is some code (either in HAL or wpilog?) that has a different set of functions/signatures depending on whether you're compiling with C or C++, so the autogenerated code is wrong when `__cplusplus` isn't specified.